File: cmd/ambience/web/effects/windmill.js
'use strict';
(function (api) {
	const { makeRNG, jitterInt, clamp01, hslToRGB, positiveMod } = api._helpers;

	const DEFAULTS = {
		intro_dur: 45,
		intro_turn: 0.12,
		ending_dur: 60,
		ending_linger: 20,
		ending_turn: 0.05,
		turn_speed: 0.08,
		blade_len: 14,
		blade_width: 1.8,
		tower_height: 20,
		tower_width: 6,
		horizon: 0.72,
		glow: 0.18,
		hue: 28,
		hue_sp: 18,
		sat: 0.42,
		lmin: 0.18,
		lmax: 0.82,
		gust_p: 0,
		lull_p: 0,
		gust_dur: 50,
		gust_mult: 1.9,
		lull_dur: 72,
		lull_mult: 0.45,
	};

	function applyDefaults(cfg) {
		const c = Object.assign({}, DEFAULTS, cfg || {});
		if (c.intro_dur <= 0) c.intro_dur = DEFAULTS.intro_dur;
		c.intro_turn = clamp01(c.intro_turn);
		if (c.ending_dur <= 0) c.ending_dur = DEFAULTS.ending_dur;
		if (c.ending_linger < 0) c.ending_linger = 0;
		c.ending_turn = clamp01(c.ending_turn);
		if (c.turn_speed <= 0) c.turn_speed = DEFAULTS.turn_speed;
		if (c.blade_len <= 0) c.blade_len = DEFAULTS.blade_len;
		if (c.blade_width <= 0) c.blade_width = DEFAULTS.blade_width;
		if (c.tower_height <= 0) c.tower_height = DEFAULTS.tower_height;
		if (c.tower_width <= 0) c.tower_width = DEFAULTS.tower_width;
		if (c.horizon <= 0) c.horizon = DEFAULTS.horizon;
		if (c.glow <= 0) c.glow = DEFAULTS.glow;
		if (c.hue === 0) c.hue = DEFAULTS.hue;
		if (c.hue_sp < 0) c.hue_sp = 0;
		if (c.sat <= 0) c.sat = DEFAULTS.sat;
		if (c.lmin <= 0) c.lmin = DEFAULTS.lmin;
		if (c.lmax <= 0) c.lmax = DEFAULTS.lmax;
		if (c.lmax < c.lmin) [c.lmin, c.lmax] = [c.lmax, c.lmin];
		if (c.gust_dur <= 0) c.gust_dur = DEFAULTS.gust_dur;
		if (c.gust_mult <= 0) c.gust_mult = DEFAULTS.gust_mult;
		if (c.lull_dur <= 0) c.lull_dur = DEFAULTS.lull_dur;
		if (c.lull_mult <= 0) c.lull_mult = DEFAULTS.lull_mult;
		return c;
	}

	class Windmill {
		constructor(w, h, cfg, seed) {
			this.w = w;
			this.h = h;
			this.grid = new Uint8ClampedArray(w * h * 3);
			this.seed = Number(seed || Date.now());
			this.tick = 0;
			this.timers = {};
			this.values = {};
			this.cfg = applyDefaults(cfg);
		}

		setConfig(cfg) {
			this.cfg = applyDefaults(Object.assign({}, this.cfg, cfg));
		}

		restoreSnapshot(snap) {
			const state = snap.state || snap;
			this.setConfig(snap.config || {});
			this.tick = state.tick || snap.tick || 0;
			this.timers = Object.assign({}, state.timers || {});
			this.values = Object.assign({}, state.values || {});
			if (typeof snap.seed === 'number') this.seed = snap.seed;
			if (snap.gridW > 0 && snap.gridH > 0) {
				this.w = snap.gridW;
				this.h = snap.gridH;
			}
		}

		_eventRng(salt) {
			return makeRNG(((this.seed >>> 0) ^ (((this.tick + salt) * 2654435761) >>> 0)) >>> 0);
		}

		_hash(index) {
			const x = Math.sin((this.seed * 0.000001 + index * 12.9898) * 43758.5453);
			return x - Math.floor(x);
		}

		_phaseProgress(total, left) {
			if (left <= 1 || total <= 1) return 1;
			const elapsed = total - left;
			if (elapsed <= 0) return 0;
			return clamp01(elapsed / Math.max(1, total - 1));
		}

		_fillCell(ctx, sx, sy, ceilSx, ceilSy, x, y, w, h, color, alpha) {
			ctx.fillStyle = color;
			ctx.globalAlpha = alpha == null ? 1 : alpha;
			ctx.fillRect(Math.floor(x * sx), Math.floor(y * sy), Math.max(1, Math.ceil(w * sx || ceilSx)), Math.max(1, Math.ceil(h * sy || ceilSy)));
			ctx.globalAlpha = 1;
		}

		triggerEvent(name) {
			const rng = this._eventRng(name.length + 71);
			switch (name) {
				case 'gust':
					this.timers.gust = jitterInt(rng, this.cfg.gust_dur, 0.3);
					this.values.gust_gain = this.cfg.gust_mult * (0.75 + rng() * 0.45);
					return true;
				case 'lull':
					this.timers.lull = jitterInt(rng, this.cfg.lull_dur, 0.3);
					return true;
				case 'intro':
					this.timers.gust = 0;
					this.timers.lull = 0;
					this.timers.ending = 0;
					this.values.gust_gain = 1;
					this.timers.intro = Math.max(1, Math.round(this.cfg.intro_dur));
					this.values.intro_total = this.timers.intro;
					return true;
				case 'ending':
					this.timers.intro = 0;
					this.timers.gust = 0;
					this.timers.lull = 0;
					this.values.gust_gain = 1;
					this.timers.ending = Math.max(1, Math.round(this.cfg.ending_dur + Math.max(0, this.cfg.ending_linger)));
					this.values.ending_total = this.timers.ending;
					return true;
			}
			return false;
		}

		step() {
			this.tick++;
			for (const key of Object.keys(this.timers)) {
				if (this.timers[key] > 0) this.timers[key]--;
			}
			if (!this.timers.gust || this.timers.gust <= 0) this.values.gust_gain = 1;
			api._helpers.paintProceduralGrid(this);
		}

		_rotationLevel() {
			let level = 1;
			if (this.timers.gust > 0) level *= this.values.gust_gain || this.cfg.gust_mult;
			if (this.timers.lull > 0) level *= this.cfg.lull_mult;
			if (this.timers.intro > 0) {
				const total = this.values.intro_total || this.cfg.intro_dur;
				const progress = this._phaseProgress(total, this.timers.intro);
				level *= this.cfg.intro_turn + (1 - this.cfg.intro_turn) * progress;
			}
			if (this.timers.ending > 0) {
				const total = this.values.ending_total || (this.cfg.ending_dur + this.cfg.ending_linger);
				const progress = this._phaseProgress(total, this.timers.ending);
				level *= 1 - (1 - this.cfg.ending_turn) * progress;
			}
			return Math.max(0.03, level);
		}

		render(ctx, canvasW, canvasH, opts) {
			api._helpers.renderPixelGridEffect(this, ctx, canvasW, canvasH, opts);
		}
	}

	api.presets['windmill'] = [
		{
			key: 'still-dusk',
			label: 'still dusk',
			config: {
				intro_turn: 0.08,
				ending_turn: 0.04,
				turn_speed: 0.04,
				blade_len: 12,
				blade_width: 1.6,
				tower_height: 19,
				tower_width: 5.5,
				horizon: 0.74,
				glow: 0.22,
				hue: 26,
				hue_sp: 14,
				sat: 0.38,
				lmin: 0.16,
				lmax: 0.78,
			},
		},
		{
			key: 'steady-turning',
			label: 'steady turning',
			config: {
				turn_speed: 0.08,
				blade_len: 14,
				blade_width: 1.8,
				tower_height: 20,
				tower_width: 6,
				horizon: 0.72,
				glow: 0.18,
				hue: 28,
				hue_sp: 18,
				sat: 0.42,
				lmin: 0.18,
				lmax: 0.82,
				gust_p: 0.0006,
			},
		},
		{
			key: 'windy-hill',
			label: 'windy hill',
			config: {
				turn_speed: 0.12,
				blade_len: 15,
				blade_width: 2.1,
				tower_height: 21,
				tower_width: 6.5,
				horizon: 0.7,
				glow: 0.14,
				hue: 24,
				hue_sp: 20,
				sat: 0.4,
				lmin: 0.16,
				lmax: 0.8,
				gust_p: 0.0014,
				gust_mult: 2.2,
				gust_dur: 62,
			},
		},
		{
			key: 'silhouette-mill',
			label: 'silhouette mill',
			config: {
				turn_speed: 0.06,
				blade_len: 16,
				blade_width: 1.5,
				tower_height: 23,
				tower_width: 5,
				horizon: 0.76,
				glow: 0.1,
				hue: 222,
				hue_sp: 12,
				sat: 0.22,
				lmin: 0.12,
				lmax: 0.68,
				lull_p: 0.0012,
				lull_mult: 0.38,
			},
		},
	];
	api.effects['windmill'] = Windmill;
})(window.AmbienceSim);

File: sim/windmill.go
package sim

import (
	"fmt"
	"math"
	"sync"

	"github.com/nelsong6/ambience/rngutil"
)

type WindmillConfig = ProceduralConfig
type WindmillState = ProceduralState
type WindmillSnapshot = ProceduralSnapshot
type WindmillPersistedState = ProceduralPersistedState

// Windmill is a dedicated sim type for the windmill effect.
type Windmill struct {
	mu sync.Mutex

	W, H int
	Grid [][]Pixel

	rng    *rngutil.RNG
	cfg    WindmillConfig
	tick   int
	timers map[string]int
	values map[string]float64
	log    []LogEntry
}

var windmillDefaultsLocal = WindmillConfig{
	"intro_dur":     45,
	"intro_turn":    0.12,
	"ending_dur":    60,
	"ending_linger": 20,
	"ending_turn":   0.05,
	"turn_speed":    0.08,
	"blade_len":     14.0,
	"blade_width":   1.8,
	"tower_height":  20.0,
	"tower_width":   6.0,
	"horizon":       0.72,
	"glow":          0.18,
	"hue":           28,
	"hue_sp":        18,
	"sat":           0.42,
	"lmin":          0.18,
	"lmax":          0.82,
	"gust_p":        0.0,
	"lull_p":        0.0,
	"gust_dur":      50,
	"gust_mult":     1.90,
	"lull_dur":      72,
	"lull_mult":     0.45,
}

func WindmillSchema() EffectSchema {
	return EffectSchema{
		Name: "windmill",
		Knobs: []Knob{
			{Key: "intro_dur", Label: "intro dur", Slot: SlotSpawn, Group: "introduction", Type: KnobInt, Min: 10, Max: 180, Step: 5, Default: 45, Trigger: "intro",
				Description: "Ticks spent easing the blades from stillness into a readable turn."},
			{Key: "intro_turn", Label: "intro turn", Slot: SlotSpawn, Group: "introduction", Type: KnobFloat, Min: 0.02, Max: 0.5, Step: 0.01, Default: 0.12,
				Description: "Starting fraction of the final rotation speed before the mill settles into motion."},
			{Key: "ending_dur", Label: "ending dur", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 60, Trigger: "ending",
				Description: "Ticks spent coasting the blades back down toward stillness."},
			{Key: "ending_linger", Label: "ending linger", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 0, Max: 160, Step: 5, Default: 20,
				Description: "Extra quiet ticks after the blades have mostly stopped."},
			{Key: "ending_turn", Label: "ending turn", Slot: SlotEnd, Group: "ending", Type: KnobFloat, Min: 0.01, Max: 0.35, Step: 0.01, Default: 0.05,
				Description: "Residual blade motion near the end of the outro."},
			{Key: "turn_speed", Label: "turn speed", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 0.02, Max: 0.25, Step: 0.01, Default: 0.08,
				Description: "Base blade rotation speed."},
			{Key: "blade_len", Label: "blade len", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 6, Max: 22, Step: 0.5, Default: 14,
				Description: "Length of the windmill blades."},
			{Key: "blade_width", Label: "blade width", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 0.5, Max: 4, Step: 0.1, Default: 1.8,
				Description: "Thickness of each blade arm."},
			{Key: "tower_height", Label: "tower height", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 10, Max: 30, Step: 0.5, Default: 20,
				Description: "Height of the windmill tower above the hill."},
			{Key: "tower_width", Label: "tower width", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 3, Max: 10, Step: 0.5, Default: 6,
				Description: "Width of the windmill tower silhouette."},
			{Key: "horizon", Label: "horizon", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 0.56, Max: 0.86, Step: 0.01, Default: 0.72,
				Description: "Height of the ground line and hill in frame."},
			{Key: "glow", Label: "glow", Slot: SlotLever, Group: "mill", Type: KnobFloat, Min: 0.02, Max: 0.5, Step: 0.01, Default: 0.18,
				Description: "Strength of the dusk haze and tiny warm window glow."},
			{Key: "hue", Label: "hue", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 10, Max: 240, Step: 1, Default: 28,
				Description: "Base sky hue spanning cool night blues through warm dusk."},
			{Key: "hue_sp", Label: "hue spread", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0, Max: 28, Step: 1, Default: 18,
				Description: "Variation between the upper sky and the horizon glow."},
			{Key: "sat", Label: "saturation", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.05, Max: 0.7, Step: 0.01, Default: 0.42,
				Description: "Overall sky and glow saturation."},
			{Key: "lmin", Label: "light min", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.05, Max: 0.5, Step: 0.01, Default: 0.18,
				Description: "Minimum lightness used for the upper sky and dark ground."},
			{Key: "lmax", Label: "light max", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.25, Max: 0.95, Step: 0.01, Default: 0.82,
				Description: "Maximum lightness used for the horizon and glow."},
			{Key: "gust_p", Label: "gust", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "gust",
				Description: "Per-tick chance of the blades briefly spinning faster."},
			{Key: "lull_p", Label: "lull", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "lull",
				Description: "Per-tick chance of the wind settling into a slower turn."},
			{Key: "gust_dur", Label: "gust dur", Slot: SlotEventMod, Group: "gust", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 50,
				Description: "Duration of a faster-turning gust."},
			{Key: "gust_mult", Label: "gust x", Slot: SlotEventMod, Group: "gust", Type: KnobFloat, Min: 1.05, Max: 3, Step: 0.05, Default: 1.9,
				Description: "Rotation multiplier applied during a gust."},
			{Key: "lull_dur", Label: "lull dur", Slot: SlotEventMod, Group: "lull", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 72,
				Description: "Duration of the calmer slower-turning window."},
			{Key: "lull_mult", Label: "lull x", Slot: SlotEventMod, Group: "lull", Type: KnobFloat, Min: 0.1, Max: 1, Step: 0.05, Default: 0.45,
				Description: "Rotation multiplier applied while lull is active."},
		},
	}
}

func defaultWindmillConfig() WindmillConfig { return cloneConfig(windmillDefaultsLocal) }

func mergeWindmillDefaults(cfg WindmillConfig) WindmillConfig {
	out := defaultWindmillConfig()
	for k, v := range cfg {
		out[k] = v
	}
	if out["intro_dur"] <= 0 {
		out["intro_dur"] = windmillDefaultsLocal["intro_dur"]
	}
	out["intro_turn"] = clamp01(out["intro_turn"])
	if out["ending_dur"] <= 0 {
		out["ending_dur"] = windmillDefaultsLocal["ending_dur"]
	}
	if out["ending_linger"] < 0 {
		out["ending_linger"] = 0
	}
	out["ending_turn"] = clamp01(out["ending_turn"])
	if out["turn_speed"] <= 0 {
		out["turn_speed"] = windmillDefaultsLocal["turn_speed"]
	}
	if out["blade_len"] <= 0 {
		out["blade_len"] = windmillDefaultsLocal["blade_len"]
	}
	if out["blade_width"] <= 0 {
		out["blade_width"] = windmillDefaultsLocal["blade_width"]
	}
	if out["tower_height"] <= 0 {
		out["tower_height"] = windmillDefaultsLocal["tower_height"]
	}
	if out["tower_width"] <= 0 {
		out["tower_width"] = windmillDefaultsLocal["tower_width"]
	}
	if out["horizon"] <= 0 {
		out["horizon"] = windmillDefaultsLocal["horizon"]
	}
	if out["glow"] <= 0 {
		out["glow"] = windmillDefaultsLocal["glow"]
	}
	if out["hue"] == 0 {
		out["hue"] = windmillDefaultsLocal["hue"]
	}
	if out["hue_sp"] < 0 {
		out["hue_sp"] = 0
	}
	if out["sat"] <= 0 {
		out["sat"] = windmillDefaultsLocal["sat"]
	}
	if out["lmin"] <= 0 {
		out["lmin"] = windmillDefaultsLocal["lmin"]
	}
	if out["lmax"] <= 0 {
		out["lmax"] = windmillDefaultsLocal["lmax"]
	}
	if out["lmax"] < out["lmin"] {
		out["lmin"], out["lmax"] = out["lmax"], out["lmin"]
	}
	if out["gust_dur"] <= 0 {
		out["gust_dur"] = windmillDefaultsLocal["gust_dur"]
	}
	if out["gust_mult"] <= 0 {
		out["gust_mult"] = windmillDefaultsLocal["gust_mult"]
	}
	if out["lull_dur"] <= 0 {
		out["lull_dur"] = windmillDefaultsLocal["lull_dur"]
	}
	if out["lull_mult"] <= 0 {
		out["lull_mult"] = windmillDefaultsLocal["lull_mult"]
	}
	return out
}

func NewWindmill(w, h int, seed int64, cfg WindmillConfig) *Windmill {
	grid := make([][]Pixel, h)
	for i := range grid {
		grid[i] = make([]Pixel, w)
	}
	return &Windmill{
		W:      w,
		H:      h,
		Grid:   grid,
		rng:    rngutil.New(seed),
		cfg:    mergeWindmillDefaults(cfg),
		timers: make(map[string]int),
		values: make(map[string]float64),
	}
}

func (w *Windmill) Resize(width, height int) {
	if width <= 0 || height <= 0 {
		return
	}
	w.mu.Lock()
	defer w.mu.Unlock()
	if width == w.W && height == w.H {
		return
	}
	w.W = width
	w.H = height
	w.Grid = make([][]Pixel, height)
	for i := range w.Grid {
		w.Grid[i] = make([]Pixel, width)
	}
}

func (w *Windmill) SetConfig(cfg WindmillConfig) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.cfg = mergeWindmillDefaults(cfg)
}

func (w *Windmill) EffectiveConfig() WindmillConfig {
	w.mu.Lock()
	defer w.mu.Unlock()
	return cloneConfig(w.cfg)
}

func (w *Windmill) Snapshot() WindmillSnapshot {
	w.mu.Lock()
	defer w.mu.Unlock()
	return WindmillSnapshot{ProceduralState: w.snapshotStateLocked()}
}

func (w *Windmill) RestoreSnapshot(snap WindmillSnapshot) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.restoreStateLocked(snap.ProceduralState)
}

func (w *Windmill) SnapshotPersistedState() WindmillPersistedState {
	w.mu.Lock()
	defer w.mu.Unlock()
	return WindmillPersistedState{
		ProceduralState: w.snapshotStateLocked(),
		RNGState:        w.rng.State(),
	}
}

func (w *Windmill) RestorePersistedState(ps WindmillPersistedState) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.restoreStateLocked(ps.ProceduralState)
	if ps.RNGState != 0 {
		w.rng.SetState(ps.RNGState)
	}
}

func (w *Windmill) snapshotStateLocked() ProceduralState {
	return ProceduralState{
		Tick:   w.tick,
		Timers: cloneTimerMap(w.timers),
		Values: cloneValueMap(w.values),
	}
}

func (w *Windmill) restoreStateLocked(state ProceduralState) {
	w.tick = state.Tick
	w.timers = cloneTimerMap(state.Timers)
	if w.timers == nil {
		w.timers = make(map[string]int)
	}
	w.values = cloneValueMap(state.Values)
	if w.values == nil {
		w.values = make(map[string]float64)
	}
}

func (w *Windmill) CurrentTick() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.tick
}

func (w *Windmill) PerturbRNG(delta int64) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.rng.Mix(delta)
}

func (w *Windmill) DrainLog() []LogEntry {
	w.mu.Lock()
	defer w.mu.Unlock()
	if len(w.log) == 0 {
		return nil
	}
	out := w.log
	w.log = nil
	return out
}

func (w *Windmill) appendLog(kind, desc string) {
	w.log = append(w.log, LogEntry{Tick: w.tick, Type: kind, Desc: desc})
	if len(w.log) > 200 {
		w.log = w.log[len(w.log)-200:]
	}
}

func (w *Windmill) intCfg(key string) int {
	return int(math.Round(w.cfg[key]))
}

func (w *Windmill) TriggerEvent(name string) bool {
	w.mu.Lock()
	defer w.mu.Unlock()
	switch name {
	case "gust":
		w.startGustLocked("triggered")
	case "lull":
		w.startLullLocked("triggered")
	case "intro":
		w.startIntroLocked()
		w.appendLog("intro", fmt.Sprintf("started (dur=%d, turn=%.2f)", w.timers["intro"], w.cfg["intro_turn"]))
	case "ending":
		w.startEndingLocked()
		w.appendLog("ending", fmt.Sprintf("started (fade=%d, linger=%d)", w.intCfg("ending_dur"), w.intCfg("ending_linger")))
	default:
		return false
	}
	return true
}

func (w *Windmill) Step() {
	w.mu.Lock()
	defer w.mu.Unlock()

	w.tick++
	for key, value := range w.timers {
		if value > 0 {
			w.timers[key] = value - 1
		}
	}
	w.stepLocked()
}

func (w *Windmill) startGustLocked(verb string) {
	w.timers["gust"] = jitterInt(w.rng, w.intCfg("gust_dur"), 0.3)
	w.values["gust_gain"] = w.cfg["gust_mult"] * (0.75 + w.rng.Float64()*0.45)
	w.appendLog("gust", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, w.timers["gust"], w.values["gust_gain"]))
}

func (w *Windmill) startLullLocked(verb string) {
	w.timers["lull"] = jitterInt(w.rng, w.intCfg("lull_dur"), 0.3)
	w.appendLog("lull", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, w.timers["lull"], w.cfg["lull_mult"]))
}

func (w *Windmill) startIntroLocked() {
	w.timers["gust"] = 0
	w.timers["lull"] = 0
	w.timers["ending"] = 0
	w.values["gust_gain"] = 1
	w.timers["intro"] = w.intCfg("intro_dur")
	w.values["intro_total"] = float64(w.timers["intro"])
}

func (w *Windmill) startEndingLocked() {
	w.timers["intro"] = 0
	w.timers["gust"] = 0
	w.timers["lull"] = 0
	w.values["gust_gain"] = 1
	endingTotal := w.intCfg("ending_dur") + max(0, w.intCfg("ending_linger"))
	if endingTotal < 1 {
		endingTotal = max(1, w.intCfg("ending_dur"))
	}
	w.timers["ending"] = endingTotal
	w.values["ending_total"] = float64(endingTotal)
}

func (w *Windmill) stepLocked() {
	if w.timers["gust"] <= 0 {
		w.values["gust_gain"] = 1
	}
	if w.timers["intro"] <= 0 {
		delete(w.values, "intro_total")
	}
	if w.timers["ending"] <= 0 {
		delete(w.values, "ending_total")
	}
	if w.timers["intro"] > 0 || w.timers["ending"] > 0 {
		return
	}
	if w.timers["gust"] <= 0 && w.cfg["gust_p"] > 0 && w.rng.Float64() < w.cfg["gust_p"] {
		w.startGustLocked("started")
	}
	if w.timers["lull"] <= 0 && w.cfg["lull_p"] > 0 && w.rng.Float64() < w.cfg["lull_p"] {
		w.startLullLocked("started")
	}
}
